### PR TITLE
Upgrade Solidity to version 5, Truffle to version 5, decode return value of onPurchase() callback

### PR DESCRIPTION
**Pull request by isentropy**
Truffle 5 allows you set the solc version in truffle-config.js, unlike version 4. Solidity 5 allows you to decode return value and catch reverts from call().

There were a number of changes required to get this working:
- upgrade solidity syntax to 5: mostly adding keywords and using address(number)
- fix the way return values are read in truffle. added function to testHelpers.js
- Truffle5 is more particular about types. Need to pass byte32 explicitly instead of string for ex
- upgrade web3 to 1.2. The 1.0-beta version caused tests to fail mysteriously with RPC timeout.
- lots of little changes to method and object names

I removed the MockUniswap contracts because we don't use them. When testing we use the real contracts.

### CODE REVIEW

File: contracts/Marketplace.sol [hunk 1]
```diff
@@ -320,21 +323,31 @@ contract Marketplace is Ownable, IMarketplace2 {
...
-        // Solidity 4:
-        // 0x91517bdd = keccak256("onPurchase(bytes32,address,uint256,uint256)")
-        // this call returns true if beneficiary is a PurchaseListener, return value is ignored
-        p.beneficiary.call(0x91517bdd, productId, subscriber, oldSub.endTimestamp, price);
+        
+        uint256 codeSize;
+        address addr = p.beneficiary;
+        assembly { codeSize := extcodesize(addr) }  // solhint-disable-line no-inline-assembly
+        if (codeSize > 0) {
+            (bool success, bytes memory returnData) = p.beneficiary.call(abi.encodeWithSignature("onPurchase(bytes32,address,uint256,uint256,uint256)",
```
Comment (isentropy) [inline contracts/Marketplace.sol]: need to check if address is a contract here. I noticed that call() returns success = true when address is a wallet.

File: contracts/Marketplace.sol [hunk 2]
```diff
@@ -320,21 +323,31 @@ contract Marketplace is Ownable, IMarketplace2 {
...
+                productId, subscriber, oldSub.endTimestamp, price, fee));
+        
+            if(success){
+                (bool accepted) = abi.decode(returnData, (bool));
+                require(accepted, "error_rejectedBySeller");
+            }
```
Comment (isentropy) [inline contracts/Marketplace.sol]: see added test for onPurchase() rejection in Marketplace.js

File: .vscode/launch.json [hunk 3]
```diff
@@ -13,6 +13,9 @@
                 "test",
                 "${file}"
             ],
+            "skipFiles": [
+                "<node_internals>/**"
+            ],
```
Comment (isentropy) [inline .vscode/launch.json]: this helps with debugging. when you pause, debugger doesn't break on node internal code

File: contracts/BancorConverter.sol [hunk 4]
```diff
@@ -1,1462 +0,0 @@
-// solhint-disable
```
Comment (isentropy) [inline contracts/BancorConverter.sol]: BancorConverter isn't used

File: contracts/deployed/Marketplace-20180425-0xa10151d088f6f2705a05d6c83719e99e079a61c1.sol [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
-pragma solidity 0.4.25;
+pragma solidity ^0.5.16;
 
 import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 
-import "../PurchaseListener.sol";
 import "../Ownable.sol";
 
+interface PurchaseListener {
```
Comment (isentropy) [inline contracts/deployed/Marketplace-20180425-0xa10151d088f6f2705a05d6c83719e99e079a61c1.sol]: inlined the old PurchaseListener interface

File: test/Marketplace.js [hunk 6]
```diff
@@ -27,10 +28,12 @@ contract("Marketplace2", accounts => {
     describe("Creating, deleting products in market1 and market2", () => {
         //product created in market1
         let id1 = "test1"
+        let id1bytes = web3.utils.padLeft(web3.utils.asciiToHex(id1), 64)
```
Comment (isentropy) [inline test/Marketplace.js]: Truffle5 doesn't accept text strings as bytes32. need to convert to bytes string

File: test/Ownable.js [hunk 7]
```diff
@@ -25,15 +25,15 @@ contract("Ownable", (accounts) => {
     })
 
     it("should prevent to claimOwnership from not-pendingOwner", async () => {
-        assertFails(ownable.claimOwnership({ from: accounts[2] }), "onlyPendingOwner")
+        assertFails(ownable.claimOwnership({ from: accounts[2] }), "Returned error: VM Exception while processing transaction: revert onlyPendingOwner -- Reason given: onlyPendingOwner.")
```
Comment (jtakalai) [inline test/Ownable.js]: eww, here it would be better to change `assertFails` to accept a regex, then match using that

Comment (jtakalai) [reply test/Ownable.js]: like, if regex, try match; if string, do like now

Comment (jtakalai) [reply test/Ownable.js]: ok found it's fixed in monoplasma, so importing the fix from there

**Comment by jtakalai** [comment]
Doesn't Travis test this repo?

**Pull request merged**